### PR TITLE
Tighten up Content-Security-Policy

### PR DESCRIPTION
**Pull request by davidmehren**
### Component/Part
Content Security Policy

### Description
This PR
- tightens the default CSP by setting `default-src` to `none` and then only allowing things explicitly.
- introduces a config option to disallow iframing HedgeDoc in other pages
- introduces a config option to disallow embedding PDFs in notes

When both options are disabled, CSP Scanner is pretty happy and gives us an [`A` score](https://cspscanner.com/?q=default-src+%27none%27%3Bbase-uri+%27self%27%3Bconnect-src+%27self%27%3Bfont-src+%27self%27%3Bmanifest-src+%27self%27%3Bframe-src+%27self%27+https%3A%2F%2Fplayer.vimeo.com+https%3A%2F%2Fwww.slideshare.net%2Fslideshow%2Fembed_code%2Fkey%2F+https%3A%2F%2Fwww.youtube.com%3Bimg-src+*%3Bscript-src+https%3A%2F%2Fhedgedoc.localhost%2Fbuild%2F+https%3A%2F%2Fhedgedoc.localhost%2Fjs%2F+https%3A%2F%2Fhedgedoc.localhost%2Fconfig+https%3A%2F%2Fgist.github.com%2F+https%3A%2F%2Fvimeo.com%2Fapi%2Foembed.json+https%3A%2F%2Fwww.slideshare.net%2Fapi%2Foembed%2F2+%27unsafe-inline%27+%27nonce-10deb111-89ad-44d3-9917-f9357a7191ec%27+%27sha256-81acLZNZISnyGYZrSuoYhpzwDTTxi7vC1YM4uNxqWaM%3D%27%3Bstyle-src+https%3A%2F%2Fhedgedoc.localhost%2Fbuild%2F+https%3A%2F%2Fhedgedoc.localhost%2Fcss%2F+%27unsafe-inline%27+https%3A%2F%2Fgithub.githubassets.com%3Bframe-ancestors+%27self%27%3Bform-action+%27self%27%3Bupgrade-insecure-requests)

By default, both options are enabled for backwards-compatibility, CSP Scanner [gives us a B](https://cspscanner.com/?q=default-src+%27none%27%3Bbase-uri+%27self%27%3Bconnect-src+%27self%27%3Bfont-src+%27self%27%3Bmanifest-src+%27self%27%3Bframe-src+%27self%27+https%3A%2F%2Fplayer.vimeo.com+https%3A%2F%2Fwww.slideshare.net%2Fslideshow%2Fembed_code%2Fkey%2F+https%3A%2F%2Fwww.youtube.com+*%3Bimg-src+*%3Bscript-src+https%3A%2F%2Fhedgedoc.localhost%2Fbuild%2F+https%3A%2F%2Fhedgedoc.localhost%2Fjs%2F+https%3A%2F%2Fhedgedoc.localhost%2Fconfig+https%3A%2F%2Fgist.github.com%2F+https%3A%2F%2Fvimeo.com%2Fapi%2Foembed.json+https%3A%2F%2Fwww.slideshare.net%2Fapi%2Foembed%2F2+%27unsafe-inline%27+%27nonce-f813cdaa-c3c5-4d78-bef9-795d92c9e0c7%27+%27sha256-81acLZNZISnyGYZrSuoYhpzwDTTxi7vC1YM4uNxqWaM%3D%27%3Bstyle-src+https%3A%2F%2Fhedgedoc.localhost%2Fbuild%2F+https%3A%2F%2Fhedgedoc.localhost%2Fcss%2F+%27unsafe-inline%27+https%3A%2F%2Fgithub.githubassets.com%3Bform-action+%27self%27%3Bobject-src+*%3Bupgrade-insecure-requests)

### Steps

- [x] Added implementation
- [ ] Added / updated tests
- [x] Added / updated documentation
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### CODE REVIEW

File: lib/csp.js [hunk 1]
```diff
@@ -4,15 +4,25 @@ const { v4: uuidv4 } = require('uuid')
...
+  frameSrc: ['\'self\'', 'https://player.vimeo.com', 'https://www.slideshare.net/slideshow/embed_code/key/', 'https://www.youtube.com'],
+  imgSrc: ['*'], // we allow using arbitrary images
+  scriptSrc: [
+    config.serverURL + '/build/',
+    config.serverURL + '/js/',
+    config.serverURL + '/config',
+    'https://gist.github.com/',
+    'https://vimeo.com/api/oembed.json',
+    'https://www.slideshare.net/api/oembed/2',
+    '\'unsafe-inline\'' // this is ignored by browsers supporting nonces/hashes
```
Comment (SISheogorath) [inline lib/csp.js:21]: Uhm, this commend is not correct. `unsafe-inline` is not ignored, if unsafe-inline is there, one can use inline scripts.

> Inline script blocks MUST pass through § 4.2.3 Should element’s inline type behavior be blocked by Content Security Policy?. Their behavior will be blocked unless every policy allows inline script, either implicitly by not specifying a script-src (or default-src) directive, or explicitly, by specifying "unsafe-inline", a nonce-source or a hash-source that matches the inline block.

https://www.w3.org/TR/CSP3/#directive-script-src

**Review by SISheogorath — CHANGES_REQUESTED** [review]

File: lib/csp.js (hunk 1 above)
Comment (davidmehren) [reply lib/csp.js:21]: MDN writes:
>  Specifying nonce makes a modern browser ignore 'unsafe-inline' which could still be set for older browsers without nonce support. 
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src

The linked CSP Scanner also does not complain about the presence of `unsafe-inline`:
> 'unsafe-inline' is ignored when a nonce or a hash is present on modern browsers, and serves for graceful degradation on legacy browsers.

Comment (SISheogorath) [reply lib/csp.js:21]: Well, I guess then I just learned something 👍🏻 

And apparently the standards are outdated again, one has to love this fast moving web.

**Review by SISheogorath — APPROVED** [review]

**Comment by davidmehren** [comment]
Now with green tests!

File: docs/content/configuration.md [hunk 2]
```diff
@@ -85,6 +85,8 @@ these are rarely used for various reasons.
...
 | `csp.upgradeInsecureRequests` |                                | **`auto`** or `true` or `false`                                                           | By default (`auto`), insecure (HTTP) requests are upgraded to HTTPS via CSP if `useSSL` is on. To change this behaviour, set to either `true` or `false`.                 |
 | `csp.reportUri`               | `CMD_CSP_REPORTURI`            | **`undefined`**, `https://<someid>.report-uri.com/r/d/csp/enforce`                        | Allows to add a URL for CSP reports in case of violations.                                                                                                                |
+| `csp.allowFraming`            | `CMD_CSP_ALLOW_FRAMING`        | **`true`** or `false`                                                                     | Disable to disallow framing of the instance. We **strongly recommend disabling** this option, as it increases the attack surface of XSS attacks.                          |
```
Comment (InnayTool) [inline docs/content/configuration.md:88]: Maybe we should explain what framing is.

Comment (ErikMichelson) [reply docs/content/configuration.md:88]: Maybe just say "embedding via iframe"?

Comment (davidmehren) [reply docs/content/configuration.md]: done

**Review by ErikMichelson — CHANGES_REQUESTED** [review]
This tightening breaks slideshare on Chrome.

```
jquery.js:10185 Refused to load the script 'http://www.slideshare.net/api/oembed/2?url=http://www.slideshare.net/briansolis/26-disruptive-technology-trends-2016-2018-56796196&format=json&callback=jQuery36008752817940203228_1628541591964&_=1628541591965' because it violates the following Content Security Policy directive: "script-src http://localhost:3000/build/ http://localhost:3000/js/ http://localhost:3000/config https://gist.github.com/ https://vimeo.com/api/oembed.json https://www.slideshare.net/api/oembed/2 'unsafe-inline' 'nonce-c14f1eb8-61a0-47e5-a4f1-5d41be287756' 'sha256-81acLZNZISnyGYZrSuoYhpzwDTTxi7vC1YM4uNxqWaM='". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
```

**Review by InnayTool — APPROVED** [review]

**Review by ErikMichelson — APPROVED** [review]

**Pull request merged**